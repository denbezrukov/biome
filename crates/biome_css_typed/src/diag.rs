use biome_diagnostics::{Diagnostic, Location, Severity, console::fmt::Formatter};
use biome_text_size::TextRange;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CssPropertyDiagnostic {
    pub kind: CssPropertyDiagnosticKind,
    pub range: Option<TextRange>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CssPropertyDiagnosticKind {
    UnknownProperty,
    InvalidValue,
    UnexpectedToken,
    DuplicateComponent,
}

impl CssPropertyDiagnosticKind {
    fn message(self) -> &'static str {
        match self {
            CssPropertyDiagnosticKind::UnknownProperty => "Unknown CSS property.",
            CssPropertyDiagnosticKind::InvalidValue => "Invalid value for CSS property.",
            CssPropertyDiagnosticKind::UnexpectedToken => "Unexpected token in CSS property value.",
            CssPropertyDiagnosticKind::DuplicateComponent => {
                "Duplicate component in CSS property value."
            }
        }
    }
}

impl Diagnostic for CssPropertyDiagnostic {
    fn severity(&self) -> Severity {
        Severity::Error
    }

    fn message(&self, fmt: &mut Formatter<'_>) -> std::io::Result<()> {
        fmt.write_str(self.kind.message())
    }

    fn description(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.write_str(self.kind.message())
    }

    fn location(&self) -> Location<'_> {
        let mut builder = Location::builder();
        if let Some(range) = &self.range {
            builder = builder.span(range);
        }
        builder.build()
    }
}
