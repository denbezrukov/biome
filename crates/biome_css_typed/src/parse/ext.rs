use biome_css_syntax::{AnyCssDimension, AnyCssGenericComponentValue, AnyCssValue};
use biome_rowan::AstNode;
use biome_text_size::TextRange;

#[allow(dead_code)]
pub(crate) trait CssNodeExt {
    /// True if this node is a generic delimiter.
    fn is_delim(&self) -> bool;

    /// Case-insensitive check if the identifier equals the given ASCII string.
    fn ident_eq_ascii(&self, needle: &str) -> bool;

    /// Case-insensitive check if the dimension unit equals the given ASCII string.
    fn unit_eq_ascii(&self, needle: &str) -> bool;

    /// Returns (value, is_integer) if this node is a number.
    fn number(&self) -> Option<(f64, bool)>;

    /// Returns a best-effort source range for diagnostics.
    fn range_opt(&self) -> Option<TextRange>;
}

impl CssNodeExt for AnyCssGenericComponentValue {
    fn is_delim(&self) -> bool {
        self.as_css_generic_delimiter().is_some()
    }

    fn ident_eq_ascii(&self, needle: &str) -> bool {
        match self.as_any_css_value() {
            Some(AnyCssValue::CssIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            Some(AnyCssValue::CssCustomIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            Some(AnyCssValue::CssDashedIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            _ => false,
        }
    }

    fn unit_eq_ascii(&self, needle: &str) -> bool {
        match self.as_any_css_value() {
            Some(AnyCssValue::AnyCssDimension(dimension)) => match dimension {
                AnyCssDimension::CssPercentage(percentage) => {
                    percentage.percent_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
                AnyCssDimension::CssRegularDimension(dimension) => {
                    dimension.unit_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
                AnyCssDimension::CssUnknownDimension(dimension) => {
                    dimension.unit_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
            },
            _ => false,
        }
    }

    fn number(&self) -> Option<(f64, bool)> {
        let number = self
            .as_any_css_value()
            .and_then(AnyCssValue::as_css_number)?;
        let token = number.value_token().ok()?;
        parse_number_literal(token.text_trimmed())
    }

    fn range_opt(&self) -> Option<TextRange> {
        Some(self.range())
    }
}

#[allow(dead_code)]
fn parse_number_literal(text: &str) -> Option<(f64, bool)> {
    let value = text.parse::<f64>().ok()?;
    Some((value, is_integer_literal(text)))
}

#[allow(dead_code)]
fn is_integer_literal(text: &str) -> bool {
    let without_sign = if let Some(rest) = text.strip_prefix('+') {
        rest
    } else if let Some(rest) = text.strip_prefix('-') {
        rest
    } else {
        text
    };

    !without_sign.is_empty() && without_sign.chars().all(|ch| ch.is_ascii_digit())
}
