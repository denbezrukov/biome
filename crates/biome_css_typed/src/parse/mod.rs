pub mod cursor_ext;
pub mod ext;

use std::iter::Peekable;

use biome_css_syntax::{AnyCssGenericComponentValue, CssGenericComponentValueList, CssLanguage};
use biome_rowan::{AstNodeList, AstNodeListIterator};

#[derive(Clone)]
pub struct CssGenericComponentValueCursor {
    iter: Peekable<AstNodeListIterator<CssLanguage, AnyCssGenericComponentValue>>,
}

impl CssGenericComponentValueCursor {
    pub fn new(list: &CssGenericComponentValueList) -> Self {
        Self {
            iter: list.iter().peekable(),
        }
    }

    pub fn is_eof(&self) -> bool {
        let mut iter = self.iter.clone();
        iter.next().is_none()
    }

    pub fn peek(&mut self) -> Option<&AnyCssGenericComponentValue> {
        self.iter.peek()
    }

    pub fn bump(&mut self) -> Option<AnyCssGenericComponentValue> {
        self.iter.next()
    }

    pub fn bump_non_delim(&mut self) -> Option<AnyCssGenericComponentValue> {
        while let Some(node) = self.bump() {
            if node.as_css_generic_delimiter().is_none() {
                return Some(node);
            }
        }
        None
    }

    pub fn peek_non_delim(&self) -> Option<AnyCssGenericComponentValue> {
        let mut iter = self.iter.clone();
        while let Some(node) = iter.next() {
            if node.as_css_generic_delimiter().is_none() {
                return Some(node);
            }
        }
        None
    }
}
