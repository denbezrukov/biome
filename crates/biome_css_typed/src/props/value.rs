use super::{
    box_sizing::BoxSizing, clear::Clear, float::Float, margin::MarginShorthand,
    padding::PaddingShorthand, position::Position, visibility::Visibility,
};

#[derive(Debug, Clone, PartialEq)]
pub enum TypedPropertyValue {
    Margin(MarginShorthand),
    Padding(PaddingShorthand),
    Visibility(Visibility),
    Position(Position),
    Float(Float),
    Clear(Clear),
    BoxSizing(BoxSizing),
}
