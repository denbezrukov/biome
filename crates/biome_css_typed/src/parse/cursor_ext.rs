use crate::parse::CssGenericComponentValueCursor;
use crate::parse::ext::CssNodeExt;

impl CssGenericComponentValueCursor {
    /// Advance over consecutive delimiter nodes.
    pub fn skip_delims(&mut self) {
        while let Some(node) = self.peek() {
            if node.is_delim() {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Save current cursor position for backtracking.
    pub fn checkpoint(&self) -> Self {
        self.clone()
    }

    /// Restore to a previously saved checkpoint.
    pub fn restore(&mut self, save: Self) {
        *self = save;
    }
}
