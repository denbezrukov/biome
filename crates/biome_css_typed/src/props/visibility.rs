use biome_css_syntax::CssGenericComponentValueList;
use biome_rowan::AstNode;
use biome_string_case::StrLikeExtension;

use crate::diag::{CssPropertyDiagnostic, CssPropertyDiagnosticKind};
use crate::parse::CssGenericComponentValueCursor;

use super::value::PropertyValue;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Visibility {
    Visible,
    Hidden,
    Collapse,
}

impl Visibility {
    pub fn parse(
        value: &CssGenericComponentValueList,
    ) -> Result<PropertyValue, CssPropertyDiagnostic> {
        let mut cur = CssGenericComponentValueCursor::new(value);

        let first = cur.bump_non_delim().ok_or_else(|| CssPropertyDiagnostic {
            kind: CssPropertyDiagnosticKind::InvalidValue,
            range: None,
        })?;

        let token = first
            .as_any_css_value()
            .and_then(|value| value.as_css_identifier())
            .and_then(|identifier| identifier.value_token().ok())
            .ok_or_else(|| CssPropertyDiagnostic {
                kind: CssPropertyDiagnosticKind::InvalidValue,
                range: Some(first.range()),
            })?;
        let ident = token.text_trimmed().to_ascii_lowercase_cow();

        let vis = match ident.as_ref() {
            "visible" => Visibility::Visible,
            "hidden" => Visibility::Hidden,
            "collapse" => Visibility::Collapse,
            _ => {
                return Err(CssPropertyDiagnostic {
                    kind: CssPropertyDiagnosticKind::InvalidValue,
                    range: Some(first.range()),
                });
            }
        };

        if let Some(extra) = cur.peek_non_delim() {
            return Err(CssPropertyDiagnostic {
                kind: CssPropertyDiagnosticKind::UnexpectedToken,
                range: Some(extra.range()),
            });
        }

        Ok(PropertyValue::Visibility(vis))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use biome_css_parser::{CssParserOptions, parse_css};
    use biome_rowan::AstNodeList;

    fn parse_visibility_value_list(source: &str) -> CssGenericComponentValueList {
        let css = format!("a {{ visibility: {source}; }}");
        let parse = parse_css(&css, CssParserOptions::default());
        assert!(
            !parse.has_errors(),
            "parser errors while parsing `{}`: {:?}",
            source,
            parse.diagnostics()
        );

        let root = parse.tree();
        let rule = root
            .rules()
            .iter()
            .next()
            .expect("expected at least one rule");
        let qualified_rule = rule
            .as_css_qualified_rule()
            .expect("expected a qualified rule");
        let block = qualified_rule.block().unwrap();
        let block = block
            .as_css_declaration_or_rule_block()
            .expect("expected declaration block");
        let item = block
            .items()
            .iter()
            .next()
            .expect("expected declaration item");
        let decl = item
            .as_css_declaration_with_semicolon()
            .expect("expected declaration with semicolon")
            .declaration()
            .unwrap();
        let property = decl.property().unwrap();
        let generic = property
            .as_css_generic_property()
            .expect("expected generic property");
        generic.value()
    }

    fn parse_visibility(source: &str) -> Result<PropertyValue, CssPropertyDiagnostic> {
        let list = parse_visibility_value_list(source);
        Visibility::parse(&list)
    }

    #[test]
    fn visibility_valid_values() {
        let results: Vec<_> = ["visible", "Hidden", "collapse"]
            .into_iter()
            .map(parse_visibility)
            .collect();
        insta::assert_debug_snapshot!("visibility_valid_values", results);
    }

    #[test]
    fn visibility_invalid_values() {
        let results: Vec<_> = ["", "inherit", "visible hidden", "visible, hidden"]
            .into_iter()
            .map(parse_visibility)
            .collect();
        insta::assert_debug_snapshot!("visibility_invalid_values", results);
    }
}
