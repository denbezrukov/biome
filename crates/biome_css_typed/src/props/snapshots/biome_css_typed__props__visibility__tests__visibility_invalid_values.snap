---
source: crates/biome_css_typed/src/props/visibility.rs
expression: results
---
[
    Err(
        CssPropertyDiagnostic {
            kind: InvalidValue,
            range: None,
        },
    ),
    Err(
        CssPropertyDiagnostic {
            kind: InvalidValue,
            range: Some(
                16..23,
            ),
        },
    ),
    Err(
        CssPropertyDiagnostic {
            kind: UnexpectedToken,
            range: Some(
                24..30,
            ),
        },
    ),
    Err(
        CssPropertyDiagnostic {
            kind: UnexpectedToken,
            range: Some(
                25..31,
            ),
        },
    ),
]
